### PR TITLE
[SYCL][XPTI] Enable PI calls notifications with arguments

### PR DESCRIPTION
**Pull request by alexbatashev**
This PR adds capability to capture PI calls' arguments and pass this data to XPTI subscribers. The sample pi-trace library demonstrates how one can parse this data on the subscriber side. An example usage of this utility would be:
```bash
XPTI_TRACE_ENABLE=1 XPTI_FRAMEWORK_DISPATCHER=lib/libxptifw.so XPTI_SUBSCRIBERS=lib/libpi_trace.so ./my_app
```

### CODE REVIEW

**Comment by smaslov-intel** [comment]
Thanks @alexbatashev ! This should help debugging issues at PI level. 

> XPTI_TRACE_ENABLE=1 XPTI_FRAMEWORK_DISPATCHER=lib/libxptifw.so XPTI_SUBSCRIBERS=lib/libpi_trace.so ./my_app

That's a lot!
Can we build/ship a tool that will make it easy to use this functionality out-of-box?

File: sycl/source/detail/plugin.hpp [hunk 1]
```diff
@@ -75,6 +127,10 @@ class plugin {
     // the per_instance_user_data field.
     std::string PIFnName = PiCallInfo.getFuncName();
     uint64_t CorrelationID = pi::emitFunctionBeginTrace(PIFnName.c_str());
+    auto ArgsData =
+        packCallArguments<PiApiOffset>(std::forward<ArgsT>(Args)...);
+    uint64_t CorrelationIDWithArgs = pi::emitFunctionWithArgsBeginTrace(
```
Comment (smaslov-intel) [inline sycl/source/detail/plugin.hpp:132]: Do we really need a new XPTI stream for emitting arguments? Can't we just extend existing one to always print arguments?

File: sycl/include/CL/sycl/detail/pi_api_id.hpp [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+/// \file pi_api_id.hpp
+/// This file contains mapping between PI API functions and their IDs. A hash
+/// function is used to generate IDs. The reason for usage of a hash function
+/// instead of raw values of PiApiKind is ABI stability. New functions can be
+/// added to PiApiKind, and there's no reliable way for external XPTI users to
+/// know when to update the values. Hashes are calculated from API function
```
Comment (smaslov-intel) [inline sycl/include/CL/sycl/detail/pi_api_id.hpp:14]: I don't quite understand the motivation here: the new PI API would be added to the end of PiApiKind, so existing values are not ever affected. XPTI users should do something fake for the piApiKind that it doesn't know about. And I don't see how using piApiID instead of piApuKind would be better in telling XPTI users "when to update".

Also, what makes the hash return unique values for all PI API function names?

**Comment by tovinkere** [comment]
> Thanks @alexbatashev ! This should help debugging issues at PI level.
> 
> > XPTI_TRACE_ENABLE=1 XPTI_FRAMEWORK_DISPATCHER=lib/libxptifw.so XPTI_SUBSCRIBERS=lib/libpi_trace.so ./my_app
> 
> That's a lot!
> Can we build/ship a tool that will make it easy to use this functionality out-of-box?

That's a possibility - or the construction of the tool could be hidden in a script.

File: sycl/source/detail/plugin.hpp (hunk 1 above)
Comment (tovinkere) [reply sycl/source/detail/plugin.hpp:132]: I think having a new stream for argument capture is a good thing - if people subscribe to the new stream, they explicitly acknowledge that this stream may have a higher cost and they would like the additional data. The older stream may be sufficient for perf tools who would like the overheads to be lower.

Comment (smaslov-intel) [reply sycl/source/detail/plugin.hpp:132]: I understand that, but I don't think there are currently (or ever be) users who don't want arguments because of its cost.

File: sycl/include/CL/sycl/detail/pi_api_id.hpp (hunk 2 above)
Comment (alexbatashev) [reply sycl/include/CL/sycl/detail/pi_api_id.hpp:14]: If we can guarantee, that the new values will be added to the end of the enum, I'll happily get rid of `PiApiID`.

> Also, what makes the hash return unique values for all PI API function names?

Nothing. Math tells us that the probability of hash collision is `1/m`. The uniqueness can be checked in compile time by specializing some template.

Anyway, I'm going to remove this file.

**Comment by alexbatashev** [comment]
> That's a lot!
> Can we build/ship a tool that will make it easy to use this functionality out-of-box?

I hardly find the existing pi_trace anything but an API usage example. It is less functional than the `SYCL_PI_TRACE=-1`. However, if it evolves to something useful, it's not a big problem to wrap it into some python script.

File: sycl/source/detail/plugin.hpp (hunk 1 above)
Comment (alexbatashev) [reply sycl/source/detail/plugin.hpp:131]: Do we have an agreement here? As a compromise, we could claim the old stream as deprecated and sunset it sometime later, when all users of existing API have a chance to migrate.

File: sycl/include/CL/sycl/detail/pi_args_helper.hpp [hunk 3]
```diff
@@ -0,0 +1,540 @@
...
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+/// \file pi_args_helper.hpp
+/// Utilities to enable PI call arguments packing for XPTI.
+///
+/// C++ wrapper for PI does not take real function argument type into account.
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/detail/pi_args_helper.hpp:12]: As an alternative approach we could modify `pi.def` to call `_PI_API` macro with APIs take.
Something like https://godbolt.org/z/4W1cYhhKf

Comment (alexbatashev) [reply sycl/include/CL/sycl/detail/pi_args_helper.hpp:12]: I like this idea, however in your particular example there's a bug (which was the reason for creating this file in the first place):

consider I want to wait for some events:
```
std::vector<event> es;
Plugin.call<PiApiKind::piEventsWait>(es.size(), es.data());
```

In this case C++ will deduce the template parameter as `size_t`, which is different from `pi_uint32`, used by the actual API. So, we'd store 8 bytes of data when we expect to decode 4.

As a workaround to that problem we could make specializations for packing helper struct through the same macro sorcery.

File: sycl/include/CL/sycl/detail/pi_api_id.hpp (hunk 2 above)
Comment (smaslov-intel) [reply sycl/include/CL/sycl/detail/pi_api_id.hpp]: Sounds good. Please add a comment to `PiApiKind` that the new values are to be added to the end since this is an external interface. I believe we should actually treat any external enum as such and not change the order of values.

File: sycl/include/CL/sycl/detail/pi_args_helper.hpp (hunk 3 above)
Comment (alexbatashev) [reply sycl/include/CL/sycl/detail/pi_args_helper.hpp]: Replaced this file with an extended macro in `pi.def`

File: sycl/tools/xpti_helpers/tuple_view.hpp [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace xpti_helpers {
+/// A helper class to provide tuple-like access to a contiguous chunk of memory.
+template <typename... Ts> struct tuple_view { unsigned char *data; };
+
+template <size_t I, typename T> struct tuple_view_element;
+
+template <size_t I, typename Head, typename... Rest>
+struct tuple_view_element<I, tuple_view<Head, Rest...>>
```
Comment (romanovvlad) [inline sycl/tools/xpti_helpers/tuple_view.hpp:26]: Could you please tell why `tuple_view` classes are required?

File: sycl/source/detail/plugin.hpp [hunk 5]
```diff
@@ -25,7 +25,67 @@ namespace sycl {
...
+  static void fill(unsigned char *Dst, size_t Offset, const T &&Arg,
+                   Args &&... Rest) {
+    using ArgsTuple = typename PiApiArgTuple<Kind>::type;
+    // C-style cast is required here.
+    auto RealArg = (typename std::tuple_element<Idx, ArgsTuple>::type)(Arg);
+    auto *Begin = reinterpret_cast<const unsigned char *>(&RealArg);
+    auto *End = Begin + sizeof(decltype(RealArg));
+    std::uninitialized_copy(Begin, End, Dst + Offset);
+    array_fill_helper<Kind, Idx + 1, Args...>::fill(
+        Dst, Offset + sizeof(decltype(RealArg)), std::forward<Args>(Rest)...);
```
Comment (romanovvlad) [inline sycl/source/detail/plugin.hpp:67]: It looks like we could adjust Dst pointer here instead of passing an offset argument.

File: sycl/source/detail/plugin.hpp [hunk 6]
```diff
@@ -25,7 +25,67 @@ namespace sycl {
...
+template <PiApiKind Kind, size_t Idx, typename T, typename... Args>
+struct array_fill_helper<Kind, Idx, T, Args...> {
+  static void fill(unsigned char *Dst, size_t Offset, const T &&Arg,
+                   Args &&... Rest) {
+    using ArgsTuple = typename PiApiArgTuple<Kind>::type;
+    // C-style cast is required here.
+    auto RealArg = (typename std::tuple_element<Idx, ArgsTuple>::type)(Arg);
+    auto *Begin = reinterpret_cast<const unsigned char *>(&RealArg);
+    auto *End = Begin + sizeof(decltype(RealArg));
+    std::uninitialized_copy(Begin, End, Dst + Offset);
```
Comment (romanovvlad) [inline sycl/source/detail/plugin.hpp:65]:
```suggestion
*(typename std::tuple_element<Idx, ArgsTuple>::type *)Dst = RealArg;
```
Assuming we get rid of offset arg(see next comment)

File: sycl/tools/xpti_helpers/tuple_view.hpp (hunk 4 above)
Comment (alexbatashev) [reply sycl/tools/xpti_helpers/tuple_view.hpp:26]: I don't know any other name to get a tuple-like view into raw memory. Even if reinterpret cast to `std::tuple` works, this sounds like too much UB.

Comment (romanovvlad) [reply sycl/tools/xpti_helpers/tuple_view.hpp:26]: Probably it's even worse, but anyway take a look at unpack2 function in  https://godbolt.org/z/eWPhdna4W

Comment (alexbatashev) [reply sycl/tools/xpti_helpers/tuple_view.hpp]: @romanovvlad I tried casting to `std::tuple<...> *`, but apparently it causes some problems with memory alignment. So, I've settled with your second proposal.

File: sycl/tools/xpti_helpers/pi_arguments_handler.hpp [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+template <typename TupleT, size_t... Is>
+inline auto get(char *Data, const std::index_sequence<Is...> &) {
+  // Our type should be last in Is sequence
+  using TargetType =
+      typename std::tuple_element<sizeof...(Is) - 1, TupleT>::type;
+
+  // Calculate sizeof all elements before target + target element then substract
+  // sizeof target element
+  size_t Res = (sizeof(typename std::tuple_element<Is, TupleT>::type) + ...) -
```
Comment (romanovvlad) [inline sycl/tools/xpti_helpers/pi_arguments_handler.hpp:28]:
```suggestion
  const size_t Offset = (sizeof(typename std::tuple_element<Is, TupleT>::type) + ...) -
```

File: sycl/tools/pi-trace/pi_trace.cpp [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+        tpCallback);
+    xptiRegisterCallback(
+        GStreamID, (uint16_t)xpti::trace_point_type_t::function_with_args_end,
+        tpCallback);
+
+#define _PI_API(api, ...)                                                      \
+  ArgHandler.set##_##api([](auto &&... Args) {                                 \
+    std::cout << "---> " << #api << "("                                        \
+              << "\n";                                                         \
+    sycl::detail::pi::printArgs(Args...);                                      \
```
Comment (romanovvlad) [inline sycl/tools/pi-trace/pi_trace.cpp:57]: Not sure I fully understand this code. It seems like this lambda will be called with a single argument - tuple. It doesn't seem like `printArgs` can print tuples. Am I missing something?

Also, can we have a simple test? It would help with understanding how it works.

Comment (alexbatashev) [reply sycl/tools/pi-trace/pi_trace.cpp:57]: I use [`std::apply`](https://en.cppreference.com/w/cpp/utility/apply) to unpack tuple into separate function call arguments inside `handle(void *)`.

**Review by romanovvlad — COMMENTED** [review]
LGTM, but still suggest adding a test.

**Comment by smaslov-intel** [comment]
> > That's a lot!
> > Can we build/ship a tool that will make it easy to use this functionality out-of-box?
> 
> I hardly find the existing pi_trace anything but an API usage example. It is less functional than the `SYCL_PI_TRACE=-1`. However, if it evolves to something useful, it's not a big problem to wrap it into some python script.

Why are you evolving it then, if you don't think it is useful?

File: sycl/source/detail/plugin.hpp (hunk 1 above)
Comment (smaslov-intel) [reply sycl/source/detail/plugin.hpp:135]: OK, let's keep 2

File: sycl/include/CL/sycl/detail/pi.def [hunk 9]
```diff
@@ -10,127 +10,207 @@
...
+// This list is used to define PiApiKind enum, which is part of external
+// interface. To avoid ABI breakage, please, add new entries to the end of the
+// list.
+//
 // Platform
-_PI_API(piPlatformsGet)
-_PI_API(piPlatformGetInfo)
-_PI_API(piextPlatformGetNativeHandle)
-_PI_API(piextPlatformCreateWithNativeHandle)
+_PI_API(piPlatformsGet, pi_uint32, pi_platform *, pi_uint32 *)
```
Comment (smaslov-intel) [inline sycl/include/CL/sycl/detail/pi.def:22]: Why did you need to duplicate list of arguments here? This way you need to keep more things in sync (pi.h & pi.def). Can't you take them from a function definition as it is done for SYCL_PI_TRACE?

Comment (alexbatashev) [reply sycl/include/CL/sycl/detail/pi.def:22]: @smaslov-intel as far as I understand, SYCL_PI_TRACE uses template magic to unwrap arguments. As I wrote in a previous comment, this has a major downside. If we somewhere call `Plugin.call<PiApiKind::smth>(..., vector.size(),...)`, C++ will deduce `size_t` as template parameter. The actual API call may expect `pi_uint32`, which is twice as small. This is a problem for the XPTI users, since they have no idea what data type to expect. So, we need to detect such situations on SYC RT side. Am I missing something? Is there a better way to do this?

**Comment by alexbatashev** [comment]
> > > That's a lot!
> > > Can we build/ship a tool that will make it easy to use this functionality out-of-box?
> > 
> > 
> > I hardly find the existing pi_trace anything but an API usage example. It is less functional than the `SYCL_PI_TRACE=-1`. However, if it evolves to something useful, it's not a big problem to wrap it into some python script.
> 
> Why are you evolving it then, if you don't think it is useful?

I'm not saying it's useless. It's just not mature to compete with `SYCL_PI_TRACE` (yet). So, I don't expect many users for now. In future `pi_trace` can very well grow more functionality (and replace the environment variable, if that makes sense).

File: sycl/include/CL/sycl/detail/pi.def (hunk 9 above)
Comment (smaslov-intel) [reply sycl/include/CL/sycl/detail/pi.def:22]: Why do you say that seeing "size_t" instead of "pi_uint32" is necessarily a problem?
The intent was that all the PI API arguments use explicit cast, and that cast would verify argument values at runtime, e.g. see here: https://github.com/intel/llvm/blob/14cdb5a4d385545fbff24e6c3d3172b14be4b663/sycl/plugins/level_zero/pi_level_zero.hpp#L47

Comment (alexbatashev) [reply sycl/include/CL/sycl/detail/pi.def:22]: Let's take a look at piPlatformsGet:
https://github.com/intel/llvm/blob/9d8e94d0a333853160d2104d8ce2e44b64d14caa/sycl/include/CL/sycl/detail/pi.h#L876-L878

In SYCL runtime I do something like:
```c++
std::vector<pi_platform> Platforms;
pi_uint32 numPlatforms;
Plugin.call<PiApiKind::piPlatformsGet>(0, nullptr, &numPlatforms);
Platforms.resize(numPlatforms);
Plugin.call<PiApiKind::piPlatformsGet>(Platforms.size(), Platforms.data(), nullptr);
```

In this case the inferred template parameters would be `size_t, pi_platform*, nullptr_t`. This means that we would pack more data, than the subscriber expects, and there's no way to know how much data was packed and how to convert it back to call arguments.

We do have such examples in the wild, like with 
https://github.com/intel/llvm/blob/9d8e94d0a333853160d2104d8ce2e44b64d14caa/sycl/include/CL/sycl/detail/pi.h#L1345-L1349

https://github.com/intel/llvm/blob/9d8e94d0a333853160d2104d8ce2e44b64d14caa/sycl/source/detail/scheduler/commands.cpp#L1761-L1764

And I don't see a better way do deal with it except for casting everything to real PI API types before packing the info.

Comment (smaslov-intel) [reply sycl/include/CL/sycl/detail/pi.def:22]: Can't you know the formal argument types from just PI API decltype, like here:
https://stackoverflow.com/a/7943765

Comment (alexbatashev) [reply sycl/include/CL/sycl/detail/pi.def:22]: @smaslov-intel  that's a brilliant idea, I'll update the patch

**Review by smaslov-intel — DISMISSED** [review]
LGTM, thanks

**Comment by alexbatashev** [comment]
@tovinkere could you please take a look? Your approval is required to merge this patch.

File: sycl/include/CL/sycl/detail/common.hpp [hunk 10]
```diff
@@ -31,6 +31,9 @@ namespace detail {
 constexpr const char *SYCL_STREAM_NAME = "sycl";
 // Stream name being used for traces generated from the SYCL plugin layer
 constexpr const char *SYCL_PICALL_STREAM_NAME = "sycl.pi";
+// Stream name being used for traces generated from PI calls. This stream
+// contains information about function arguments.
+constexpr const char *SYCL_PIARGCALL_STREAM_NAME = "sycl.pi.arg";
```
[the remark this answers is not in the document]
Comment (tovinkere) [reply sycl/include/CL/sycl/detail/common.hpp:36]: I strongly recommend naming the stream "sycl.pi.debug" instead of "arg".

File: xpti/include/xpti_data_types.h [hunk 11]
```diff
@@ -255,6 +269,10 @@ enum class trace_point_type_t : uint16_t {
   function_end = XPTI_TRACE_POINT_END(12),
   /// Use to notify that a new metadata entry is available for a given event
   metadata = XPTI_TRACE_POINT_BEGIN(13),
+  /// Used to trace function call begin and its arguments.
+  function_with_args_begin = XPTI_TRACE_POINT_BEGIN(14),
```
Comment (tovinkere) [inline xpti/include/xpti_data_types.h:273]: Just FYI - In the future, you don't need to extend this enum. You can extend it as a "user_defined" trace point.

**Review by tovinkere — APPROVED** [review]
LGTM, thanks.

File: sycl/source/detail/plugin.hpp [hunk 12]
```diff
@@ -75,6 +127,10 @@ class plugin {
     // the per_instance_user_data field.
     std::string PIFnName = PiCallInfo.getFuncName();
     uint64_t CorrelationID = pi::emitFunctionBeginTrace(PIFnName.c_str());
+    auto ArgsData =
+        packCallArguments<PiApiOffset>(std::forward<ArgsT>(Args)...);
+    uint64_t CorrelationIDWithArgs = pi::emitFunctionWithArgsBeginTrace(
+        PiApiID<PiApiOffset>::id, ArgsData.data());
```
[the remark this answers is not in the document]
Comment (alexbatashev) [reply sycl/source/detail/plugin.hpp]: Right now I solve this problem by using _PI_API macro and values from `pi.def` file, but I can send function name as well.

**Review by smaslov-intel — APPROVED** [review]

**Comment by alexbatashev** [comment]
@tovinkere friendly ping

**Review by tovinkere — APPROVED** [review]
LGTM

**Comment by alexbatashev** [comment]
The pre-commit failure is not related to this patch.
@bader could you please review/merge?
